assets/js/components/permissions-modal/index.js

```diff
@@ -0,0 +1,84 @@
+/**
+ * PermissionsModal component.
+ *
+ * Site Kit by Google, Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { useCallback } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import Data from 'googlesitekit-data';
+import { STORE_NAME as CORE_USER } from '../../googlesitekit/datastore/user';
+import Dialog from '../dialog';
+import Modal from '../modal';
+import { snapshotAllStores } from '../../googlesitekit/data/create-snapshot-store';
+
+const { useSelect, useDispatch, useRegistry } = Data;
+
+const PermissionsModal = () => {
+	const registry = useRegistry();
+	const permissionsError = useSelect( ( select ) => select( CORE_USER ).getPermissionScopeError() );
+	const connectURL = useSelect(
+		( select ) => select( CORE_USER ).getConnectURL( {
+			additionalScopes: permissionsError?.data?.scopes,
+			redirectURL: global.location.href,
+		} )
+	);
+	const { clearPermissionScopeError } = useDispatch( CORE_USER );
+
+	const onCancel = useCallback( () => {
+		clearPermissionScopeError();
+	}, [ clearPermissionScopeError ] );
+
+	const onConfirm = useCallback( async () => {
+		// If we have a datastores to snapshot before navigating away to the
+		// authorization page, do that first.
+		await snapshotAllStores( registry );
+
+		global.location.assign( connectURL );
+	}, [ registry, connectURL ] );
+
+	if ( ! permissionsError ) {
+		return null;
+	}
+
+	// If there aren't any scopes for us to request, there's no reason to show
+	// the modal. Log a console warning if this happens and return `null`.
+	if ( ! permissionsError?.data?.scopes?.length ) {
+		global.console.warn( 'permissionsError lacks scopes array to use for redirect, so not showing the PermissionsModal. permissionsError was:', permissionsError );
+		return null;
+	}
+
+	return (
+		<Modal>
+			<Dialog
+				title={ __( 'Additional Permissions Required', 'google-site-kit' ) }
+				subtitle={ permissionsError.message }
```

Comment:
We may want to decouple the error message from the message we display here. `error.message` is an actual error message, but what we display here is not - we "catch" the error and then display more helpful explanation/instructions here that is actionable. In the IB we were talking about using `error.data.requestScopesMessage`, so it may be more appropriate to rely on that, and if it doesn't exist, use a fallback defined here, e.g. "In order to proceed, you need to grant additional permissions." (exact wording TBD). See also https://github.com/google/site-kit-wp/pull/1582#discussion_r430800277

Reply:
See my reply to https://github.com/google/site-kit-wp/pull/1582#discussion_r430800277 - I'm fine with whatever we decide to do there.

Reply:
Let's keep as is for now.